pkg/kubectl/cmd/auth.go

```diff
@@ -0,0 +1,451 @@
+	}
+	user, ok := cfg.AuthInfos[ctx.AuthInfo]
+	if !ok {
+		return nil, fmt.Errorf("user %q does not exist in the configuration", ctx.AuthInfo)
+	}
+	return user, nil
+}
+
+func GetCurrentCluster(cfg *clientcmdapi.Config) (*clientcmdapi.Cluster, error) {
+	ctx, err := GetCurrentContext(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	if ctx.Cluster == "" {
+		return nil, errors.New("the current-context must have a cluster set")
+	}
+	cluster, ok := cfg.Clusters[ctx.Cluster]
+	if !ok {
+		return nil, fmt.Errorf("cluster %q does not exist in the configuration", ctx.Cluster)
+	}
+	return cluster, nil
+}
+
+// OIDCAuthClient is an object performs operations related to OpenID Connect Authentication against the API Server.
+type OIDCAuthClient struct {
+	server     url.URL
+	httpClient *http.Client
+}
+
+// NewOIDCAuthClient creates a new OIDCAuthClient for the given cluster.
+func NewOIDCAuthClient(cluster *clientcmdapi.Cluster) (*OIDCAuthClient, error) {
+	if cluster.Server == "" {
+		return nil, fmt.Errorf("cluster has no server set")
+	}
+
+	srvURL, err := url.Parse(cluster.Server)
+	if err != nil {
+		return nil, err
+	}
+
+	httpClient, err := HTTPClientForCluster(cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	return &OIDCAuthClient{
+		server:     *srvURL,
+		httpClient: httpClient,
+	}, nil
+}
+
+// ExchangeRefreshToken exchanges the given refresh token for an ID token.
+func (o *OIDCAuthClient) ExchangeRefreshToken(refreshToken string) (string, error) {
+	reqURL := o.server
+	reqURL.Path = oidc.PathExchangeRefreshToken
+	q := url.Values{}
+	q.Set("refresh_token", refreshToken)
+	req, err := http.NewRequest("POST", reqURL.String(),
+		bytes.NewBuffer([]byte(q.Encode())))
+	if err != nil {
+		return "", err
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	resp, err := o.httpClient.Do(req)
+	if err != nil {
+		return "", err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("Non 200 status returned from server: %v", resp.StatusCode)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("Server returned non-200 status: %d", resp.StatusCode)
+	}
+
+	raw, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+
+	q, err = url.ParseQuery(string(raw))
+	if err != nil {
+		return "", err
+	}
+
+	idToken := q.Get("id_token")
+	if idToken == "" {
+		return "", errors.New("Missing ID token from response")
+	}
+
+	return idToken, err
+}
+
+// Login opens a browser for the user to authenticate to, and uses those credentials to obtain an ID token and refresh token, which are the return values (respectively) along with an error.
+// This method also starts a web server listening on a random port on localhost
+// to facilitate communication back from the browser.
+func (o *OIDCAuthClient) Login() (string, string, error) {
+
+	// Pick a random open port to listen on.
+	listener, err := net.Listen("tcp", "127.0.0.1:0")
```

Comment:
All communication outside of the computer running `kubectl` is over https. This server runs locally, listening only on the loopback network. FWIW, Google endorses this method of transferring auth code to native clients. See:

https://developers.google.com/identity/protocols/OAuth2InstalledApp#overview

and 

https://developers.google.com/identity/protocols/OAuth2InstalledApp#choosingredirecturi